h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -71,73 +105,39 @@ private static void extractFrameKeys(final Set<Key> retainedkeys, final Frame fr
     }
   }
 
-  /**
-   * Exctracts keys a {@link Model} points to.
-   *
-   * @param retainedKeys A set of retained keys to insert the extracted {@link Model} keys to.
-   * @param model        An instance of {@link Model} to extract the keys from
-   */
-  private static void extractModelKeys(final Set<Key> retainedKeys, final Model model) {
-    Objects.requireNonNull(model);
-    if (model._parms._train != null) {
-      retainedKeys.add(model._parms._train);
-      extractFrameKeys(retainedKeys, model._parms._train.get());
-    }
-    if (model._parms._valid != null) {
-      retainedKeys.add(model._parms._valid);
-      extractFrameKeys(retainedKeys, model._parms._valid.get());
-    }
-  }
-
-  private static class ClearDKVTask extends MRTask<ClearDKVTask> {
+  private static final class NodeKeysRemovalTask extends H2O.RemoteRunnable<NodeKeysRemovalTask> {
 
-    private final Key[] _retainedKeys; // Original model & frame keys provided by the user, accompanied with extracted internal keys
+    private final Key[] _ignoredKeys;
 
-    /**
-     *
-     * @param retainedKeys Keys that are NOT deleted and will remain in DKV. Only Model keys and Frame keys are accepted
-     */
-    public ClearDKVTask(Key[] retainedKeys) {
-      _retainedKeys = retainedKeys;
+    private NodeKeysRemovalTask(final Key[] retainedKeys) {
+      _ignoredKeys = retainedKeys;
     }
 
     @Override
-    protected void setupLocal() {
-      final Set<Key> retainedKeys = new HashSet<>(_retainedKeys.length);
-      retainedKeys.addAll(Arrays.asList(_retainedKeys));
-
-      final Collection<Value> storeKeys = H2O.STORE.values();
-      Futures removalFutures = new Futures();
-      for (final Value value : storeKeys) {
-        if (retainedKeys.contains(value._key)) {
-          continue;
+    public void run() {
+      final Set<Key> values = H2O.localKeySet();
+      final Set<Key> ignoredSet = new HashSet<>();
+      final Futures futures = new Futures();
+
+      for (final Key ignoredKey : _ignoredKeys) {
+        ignoredSet.add(ignoredKey);
+      }
+
+      for (final Key key : values) {
+        if (ignoredSet.contains(key)) continue; // Do not perform DKV.get at all if the key is to be ignored
+        if (key.isChunkKey()) continue; // Chunks are deleted with vectors & are not Keyed
+
+        final Value value = DKV.get(key);
+        if (value == null || value.isNull()) continue;
+        if (value.isModel()) {
+          Keyed.remove(key, futures, true);
```

Comment:
Why cascade=True? This is why need to extract the frames from the models in the first place. If cascade = false then we don't need to do https://github.com/h2oai/h2o-3/pull/4311/files#diff-6d64c422677c3739279b45f7e56d72abL24

Reply:
False !